### PR TITLE
Handle invisible characters in forms

### PR DESCRIPTION
**Pull request by kosmydel**
### Details

This PR handles invisible characters in the required fields, to prevent users from creating an i.g. workspace name with only invisible characters.

We:
- remove all invisible characters, and then check if are there any visible ones left,
- then we remove invisible characters that do not break the emojis (some emojis use helping invisible characters), before sending it to the BE.

### Fixed Issues

$ https://github.com/Expensify/App/issues/23297
PROPOSAL: https://github.com/Expensify/App/issues/23297#issuecomment-1686128116

### Tests

- [x] Do the QA tests
- [x] Write unit tests to verify:
  - [x] `isEmptyString` function,
  - [x] `removeInvisibleCharacters` function.

‍
### Offline tests

N/A

### QA Steps

- [ ] Verify that no errors appear in the JS console
- [ ] Verify invisible characters. In each field/form below, do the steps from the next point.
  - [ ] Verify workspace names:
    - [ ] Go to Settings -> Workspaces‍
    - [ ] Create a New workspace or use an existing one, select it
    - [ ] With workspace selected go to Settings, type in the `Name` field
  - [ ] Go to Settings -> Profile -> Personal details -> Home address -> Check: Address line, City and other fields
  - [ ] Go to Settings -> Wallet -> Add debit card -> Check: Billing address, and other fields
  - [ ] Try to cover as many forms/fields as possible, especially the ones which are required
- [ ] In each form (field) from above verify:
  - [ ] Try to change input, to a normal one, containing normal characters (e.g. `My workspace 123`)
  - [ ] Try input with only invisible characters:
    - [ ] Try characters from [this site](https://www.editpad.org/tool/invisible-character)
    - [ ] Test [U+3164 HANGUL FILLER](https://invisible-characters.com/3164-HANGUL-FILLER.html).
    - [ ] ‍Try invisible characters used in emojis (e.g.r `\u200D`), you can use [this tool](https://r12a.github.io/app-conversion/) to copy it (paste the character in the *JS/Java/C* section, click *Convert*, and then *Copy* in *Characters* section)
  - [ ] Try input with emojis, especially tricky ones (e.g. 😶‍🌫️, 🏴󠁧󠁢󠁥󠁮󠁧󠁿),‍‍
  - [ ] Try some different characters (i.g. from other alphabets)
- [ ] Verify other forms:
  - [ ] Verify Settings -> Profile -> Display Name (should do both First name and Last name):
    - [ ] Try to set the name to a normal string.
    - [ ] Try to set the name with Emojis.
    - [ ] Try to set the name to empty one.
    - [ ] Try to set the name to one with only invisible characters (e.g. `\u200D`) (should be able to save, but the name will be treated as empty)
  - [ ] Click the "Plus" button -> Assign task -> Title:
    - [ ] Do the same check as above.
  - [ ] Do the same with: Click "Plus" button -> Save the world -> I know a teacher -> All fields
  - [ ] Do the same with: Click "Plus" button -> Save the world -> I am a teacher -> All fields
  - [ ] Do the same with: Settings -> Profile -> Personal details -> Legal Name
  - [ ] Do the same with: Settings -> Profile -> Personal details -> Home address
  - [ ] Enable betas and do the same with: Fab -> Send message -> Room -> Room name
- [ ] Verify all other Forms, if you find any more.
- [ ] Go to Workspace settings -> Notes -> Edit note
- [ ] Double test recently migrated forms (i.g. [this PR](https://github.com/Expensify/App/pull/29098/files))

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`‍
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else‍‍‍
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/104823336/af4e03e4-1531-46e9-9c30-006d80879acd

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/104823336/9a1b6963-79bf-4f13-b260-28f4e723cdfd

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/104823336/0407493d-52e2-4aef-938c-b4976c27adc3

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/104823336/9812b2f2-1638-4391-a4e8-f2f13ce4e714

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/104823336/e6613d31-2e7c-4c93-a9c6-c94e38e357fb

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/104823336/6d1a0f69-4803-48cb-9789-34f7c770ead7

</details>

### CODE REVIEW

**Comment by situchan** [comment]
@kosmydel please update QA step in more detail.
We should test fix in all possible pages - new/edit room, edit workspace name, first name, last name, etc where validations were not that strict originally.

**Comment by kosmydel** [comment]
> @kosmydel please update QA step in more detail. We should test fix in all possible pages - new/edit room, edit workspace name, first name, last name, etc where validations were not that strict originally.

Merged with the main & slightly updated the PR. 
I've also updated the QA steps, but I couldn't find a way to create/edit a room.

**Comment by kosmydel** [comment]
> To create room, enable beta in Permissions.js and then go to FAB > Send message > Room tab

cc @situchan done, thanks!

**Comment by situchan** [comment]
Server is down - https://expensify.slack.com/archives/C01GTK53T8Q/p1695314249132269
Will checklist once it's back

**Comment by situchan** [comment]
@kosmydel please fix conflict

**Comment by kosmydel** [comment]
Hey @situchan. 

Since [this PR](https://github.com/Expensify/App/pull/27025) related to the Form component was merged recently, I have to apply my solution to the refactored forms. Unfortunately, I have other issues with higher priority. I will try to have a look at this by the end of this week.

**Comment by situchan** [comment]
No problem. Thanks for the update

**Comment by kosmydel** [comment]
Hey @situchan, there is a problem with implementing it in the new Form component. It doesn't implement full validation yet.

---

[Here is how the validation looks right now.](https://github.com/Expensify/App/blob/28b30fc102bd41ee6961785bc2dbc22aae52e422/src/components/Form/FormProvider.js#L107-L114)
[Here is how it looks in the previous version.](https://github.com/Expensify/App/blob/28b30fc102bd41ee6961785bc2dbc22aae52e422/src/components/Form.js#L122)

I've talked with @kowczarz and he is preparing an update to the new component form. I think that we should wait for his changes to be merged.

---

Alternatively, we can process with only the old Form component and create a new PR with changes to the new one. The problem is, that some of the components are already migrated to the new Form (e.g. `DisplayNamePage.js`), and wouldn't include changes from this PR.

**Comment by situchan** [comment]
I think we can still wait as this is not critical bug.
cc: @hayata-suenaga

**Comment by hayata-suenaga** [comment]
@kosmydel is there any one GH issue or PR against which I can hold this PR?

**Comment by kosmydel** [comment]
Hey @hayata-suenaga 
I don't think there is a GH issue, but [here is a draft of the PR](https://github.com/Expensify/App/pull/28821) fixing it.

**Comment by kosmydel** [comment]
Hey @situchan 
I've resolved conflicts and applied solutions to the migrated Form component.

I also found one invisible character, that is not in the Unicode format and control groups. I've added a case for it.

**Review by hayata-suenaga — APPROVED** [review]

**Comment by hayata-suenaga** [comment]
Situchan is going to complete the checklist once the server is back

**Comment by situchan** [comment]
Reviewing today. @kosmydel please pull main for safety. 1.6k commits behind

**Comment by kosmydel** [comment]
I've pulled the main and quickly tested it. @situchan

**Comment by situchan** [comment]
@kosmydel please fix conflict when you get a chance

**Review by hayata-suenaga — DISMISSED** [review]

**Comment by hayata-suenaga** [comment]
@kosmydel please fix the conflict when you see this message

**Comment by kosmydel** [comment]
Hey @situchan, I've resolved conflicts. Sorry for the delay, I was OOO on Tuesday, and yesterday we had public holidays in Poland.

**Comment by hayata-suenaga** [comment]
@situchan please review this when you got a minute 🙇

File: src/components/Form.js
Comment (situchan) [inline src/components/Form.js:130]: `useMemo` is more accurate here.
Actually, we can take this function out of component lifecycle as it requires no dependency.

File: src/libs/isEmptyString.ts
Comment (situchan) [inline src/libs/isEmptyString.ts:6]: Why not define this function in StringUtils?

File: src/libs/removeInvisibleCharacters.ts
Comment (situchan) [inline src/libs/removeInvisibleCharacters.ts:6]: Why not define this function in StringUtils?

**Review by situchan — COMMENTED** [review]
Overall looks good. Some minor code dry feedback

**Comment by situchan** [comment]
Performance Tests failing. I believe not caused by this PR but we should still wait until fixed in main and merge again

**Comment by kosmydel** [comment]
Hey, sorry for the delay, I was OOO. I think it is ready for another review. @situchan

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>


https://github.com/Expensify/App/assets/108292595/9ca13211-6e9d-49e0-81d2-729cd3bf5e3c



</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/108292595/4e8ea660-c536-465d-9c45-49e540f80778



</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/108292595/c3f9c9fe-4d22-4c17-9d36-be58101c4751



</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/Expensify/App/assets/108292595/f92d0abd-9f29-4410-95f2-c7026dc8345b



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/108292595/396c7fee-43de-4555-8944-ac0d259f95b8


https://github.com/Expensify/App/assets/108292595/34a7fb17-fec6-4a70-9f85-9b16cac9b7a7



</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/Expensify/App/assets/108292595/a7456a20-0d77-48e7-893b-26682187a27a



</details>

**Review by situchan — APPROVED** [review]
Great work!

**Review by hayata-suenaga — APPROVED** [review]
👍

**Pull request merged**
